sdks/python/apache_beam/transforms/util_test.py

```diff
@@ -423,6 +432,50 @@ def test_reshuffle_streaming_global_window(self):
                 label='after reshuffle')
     pipeline.run()
 
+  def test_reshuffle_preserves_timestamps(self):
+    pipeline = TestPipeline()
+
+    # Create a PCollection and assign each element with a different timestamp.
+    before_reshuffle = (pipeline
+                        | "Four elements" >> beam.Create([
+                            {'name': 'foo', 'timestamp': MIN_TIMESTAMP},
+                            {'name': 'foo', 'timestamp': 0},
+                            {'name': 'bar', 'timestamp': 33},
+                            {'name': 'bar', 'timestamp': MAX_TIMESTAMP},
+                        ])
+                        | "With timestamp" >> beam.Map(
+                            lambda element: beam.window.TimestampedValue(element, element['timestamp']))
+    )
+
+    # Reshuffle the PCollection above and assign the timestamp of an element to that element again.
+    after_reshuffle = (before_reshuffle
+                       | "Reshuffle" >> beam.Reshuffle()
+                       | "With timestamps again" >> beam.ParDo(AddTimestamp())
+    )
+
+    # Combine each element in before_reshuffle with its timestamp.
+    formatted_before_reshuffle = (before_reshuffle
+                                  | "Get before_reshuffle timestamp" >> beam.ParDo(GetTimestamp())
+    )
+
+    # Combine each element in after_reshuffle with its timestamp.
+    formatted_after_reshuffle = (after_reshuffle
+                                 | "Get after_reshuffle timestamp" >> beam.ParDo(GetTimestamp())
+    )
+
+    expected_data = ["Timestamp(-9223372036854.775000) - foo",
+                     "Timestamp(0) - foo",
+                     "Timestamp(33) - bar",
+                     "Timestamp(9223372036854.775000) - bar"
+    ]
+
+    # Can't compare formatted_before_reshuffle and formatted_after_reshuffle directly, because they are
```

Comment:
More specifically, the equal_to must not be a (deferred) PCollection, but an concrete set of values.